src/python/pants/backend/python/rules/run_setup_py.py

```diff
@@ -0,0 +1,532 @@
+    console_scripts = entry_points['console_scripts'] = entry_points.get('console_scripts', [])
+    console_scripts.append(f'{key}={binary.adaptor.entry_point}')
+
+  # Generate the setup script.
+  setup_py_content = SETUP_BOILERPLATE.format(
+    target_address_spec=ht.address.reference(),
+    setup_kwargs_str=distutils_repr(setup_kwargs)
+  ).encode()
+  setup_py_digest = await Get[Digest](
+    InputFilesContent([FileContent('setup.py', setup_py_content)]))
+
+  chroot_digest = await Get[Digest](DirectoriesToMerge((src_digest, setup_py_digest)))
+  return SetupPyChroot(chroot_digest, json.dumps(setup_kwargs, sort_keys=True))
+
+
+@rule
+async def get_sources(request: SetupPySourcesRequest,
+                      source_root_config: SourceRootConfig) -> SetupPySources:
+  targets = request.hydrated_targets
+  stripped_srcs_list = await MultiGet(
+    Get[SourceRootStrippedSources](HydratedTarget, target) for target in targets)
+
+  # Create a chroot with all the sources, and any ancestor __init__.py files that might be needed
+  # for imports to work.  Note that if a repo has multiple exported targets under a single ancestor
+  # package, then that package must be a namespace package, which in Python 3 means it must not
+  # have an __init__.py. We don't validate this here, because it would require inspecting *all*
+  # targets, whether or not they are in the target set for this run - basically the entire repo.
+  # So it's the repo owners' responsibility to ensure __init__.py hygiene.
+  stripped_srcs_digests = [stripped_sources.snapshot.directory_digest
+                           for stripped_sources in stripped_srcs_list]
+  ancestor_init_pys = await Get[AncestorInitPyFiles](HydratedTargets, targets)
+  sources_digest = await Get[Digest](
+    DirectoriesToMerge(directories=tuple([*stripped_srcs_digests, *ancestor_init_pys.digests])))
+  sources_snapshot = await Get[Snapshot](Digest, sources_digest)
+  packages, namespace_packages, package_data = find_packages(
+    source_root_config.get_source_roots(), zip(targets, stripped_srcs_list), sources_snapshot)
+  return SetupPySources(digest=sources_digest, packages=packages,
+                        namespace_packages=namespace_packages, package_data=package_data)
+
+
+@rule
+async def get_ancestor_init_py(
+    targets: HydratedTargets,
+    source_root_config: SourceRootConfig
+) -> AncestorInitPyFiles:
+  """Find any ancestor __init__.py files for the given targets.
+
+  Includes sibling __init__.py files. Returns the files stripped of their source roots.
+  """
+  source_roots = source_root_config.get_source_roots()
+  # Find the ancestors of all dirs containing .py files, including those dirs themselves.
+  source_dir_ancestors: Set[Tuple[str, str]] = set()  # Items are (src_root, path incl. src_root).
+  sources_snapshots = await MultiGet(
+    Get[Snapshot](Digest, target.adaptor.sources.snapshot.directory_digest)
```

Comment:
`target.adaptor.sources.snapshot` is already a `Snapshot`, so you shouldn't need to re-load it here.

Reply:
🤦‍♂